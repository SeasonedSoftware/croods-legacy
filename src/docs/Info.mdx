---
name: Info
route: /info
order: 9
---

# Info

The Info component provides a way to fetch one resource and render a component using the resource data.
This component does a `GET` request to the `path` and calls the render function passing the resource data as the `info` parameter.

## API

| Property      | Type          | Required | Default                  | Description                                                                                                |
| ------------- | ------------- | -------- | ------------------------ | ---------------------------------------------------------------------------------------------------------- |
| id            | String/Number | true     | -                        | Ex: 1234 or '1234'                                                                                         |
| name          | String        | true     | -                        | Defines the deep properties used in the component. Ex: foo.bar                                             |
| render        | Func          | true     | -                        | A function returning a React Node. Ex: (info = {}, props) -> < JSX />                                      |
| parentId      | String/Number | false    | -                        | Parent element's Id, for more specificity in changes.                                                      |
| path          | String        | false    | `/${name}/${id}`         | Ex: /foo/bar-\_12?&=                                                                                       |
| renderError   | Func          | false    | Provider's renderError   | A function returning a React Node that will render when occurs an error. Ex: (error = {}) -> < JSX />      |
| renderLoading | Func          | false    | Provider's renderLoading | A function returning a React Node that will be rendered while wait the server response. Ex: () -> < JSX /> |

## How to use

You have to add a reducer with the same name that you will pass to Info.

```javascript
import { createReducer } from 'croods'

const reducers = {
  foobar: createReducer('foobar'),
}
```

### render

The render function receives the parameter **info** containing the response body.

## Example

You can see a working sample of this component's usage [here](https://github.com/SeasonedSoftware/croods/blob/master/example/src/colors/Info.js).
