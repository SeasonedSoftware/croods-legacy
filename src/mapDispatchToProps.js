import { bindActionCreators } from 'redux'

import createActions from './createActions'

export default (dispatch, { options, ...props }) => ({
  actions: bindActionCreators(
    createActions({ ...options, ...props }),
    dispatch,
  ),
})
