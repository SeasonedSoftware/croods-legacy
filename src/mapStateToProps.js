import find from 'lodash/find'
import get from 'lodash/get'

export default (state, { name, parentId }) => {
  const statePart = get(state, name)
  const currentState = find(statePart, block => block.parentId === parentId)
  return get(currentState, 'state') || {}
}
