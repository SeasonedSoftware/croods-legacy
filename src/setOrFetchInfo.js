import find from 'lodash/find'

export default ({ id, info, list, actions }) => {
  const item = find(
    list || [],
    item => item && item.id.toString() === id.toString(),
  )

  if (item) {
    actions.setInfo(item)
  } else if (!info || info.id.toString() !== id.toString()) {
    actions.fetchInfo(id)
  }
}
