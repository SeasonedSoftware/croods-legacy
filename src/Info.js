import { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import withOptions from './withOptions'
import mapStateToProps from './mapStateToProps'
import mapDispatchToProps from './mapDispatchToProps'
import setOrFetchInfo from './setOrFetchInfo'
import renderIfPresent from './renderIfPresent'
import renderInfoLoading from './renderInfoLoading'

class Info extends Component {
  constructor(props) {
    super(props)
    setOrFetchInfo(props)
  }

  componentDidUpdate(prevProps) {
    const { id } = this.props
    const { id: oldId } = prevProps

    if (id.toString() !== oldId.toString()) {
      setOrFetchInfo(this.props)
    }
  }

  render() {
    const { render, info, infoError, renderError } = this.props

    return (
      renderIfPresent(renderError, infoError) ||
      renderInfoLoading(this.props) ||
      render(info, this.props)
    )
  }
}

Info.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  info: PropTypes.object.isRequired,
  render: PropTypes.func.isRequired,
  parentId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  renderError: PropTypes.func,
}

// INTERNAL PROPS
// infoError: object,

export default withOptions(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(Info),
)
