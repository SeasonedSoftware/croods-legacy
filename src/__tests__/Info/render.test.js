import React from 'react'
import { connect } from 'react-redux'
import renderer from 'react-test-renderer'

import Info from '../../Info'
import mapStateToProps from '../../mapStateToProps'
import mapDispatchToProps from '../../mapDispatchToProps'
import setOrFetchInfo from '../../setOrFetchInfo'

jest.mock('../../withOptions', () => Component => Component)

jest.mock('react-redux', () => ({
  connect: jest.fn(() => Component => Component),
}))

jest.mock('../../setOrFetchInfo', () => jest.fn())

const render = jest.fn((info, props) => (
  <div info={info} {...props}>
    Info
  </div>
))

beforeEach(() => {
  render.mockClear()
  setOrFetchInfo.mockClear()
})

it('renders correctly', () => {
  const props = { id: 1, name: 'colors', render }
  const tree = renderer.create(<Info {...props} />).toJSON()

  expect(tree).toMatchSnapshot()
  expect(connect).toHaveBeenCalledWith(mapStateToProps, mapDispatchToProps)
  expect(setOrFetchInfo).toHaveBeenCalledWith(props)
})

describe('with renderError and without infoError', () => {
  it('renders Info', () => {
    const renderError = jest.fn(error => <div>Error: {error}</div>)
    const props = { render, renderError }
    const tree = renderer
      .create(<Info id={1} name="colors" {...props} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe('with renderError and infoError', () => {
  it('renders error', () => {
    const renderError = jest.fn(error => <div>Error: {error}</div>)
    const props = { render, renderError, infoError: 'foo' }
    const tree = renderer
      .create(<Info id={1} name="colors" {...props} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe('with renderLoading and fetchingInfo', () => {
  it('renders loading correctly', () => {
    const renderLoading = jest.fn(() => <div>Loading</div>)
    const props = { render, fetchingInfo: true, renderLoading }
    const tree = renderer
      .create(<Info id={1} name="colors" {...props} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
    expect(render).not.toHaveBeenCalled()
  })
})

describe('with renderLoading and info', () => {
  it('renders info correctly', () => {
    const renderLoading = jest.fn(() => <div>Loading</div>)
    const props = { render, renderLoading, info: { foo: 'bar' } }
    const tree = renderer
      .create(<Info id={1} name="colors" {...props} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe('without renderLoading and with fetchingInfo', () => {
  it('renders info correctly', () => {
    const props = { render, fetchingInfo: true }
    const tree = renderer
      .create(<Info id={1} name="colors" {...props} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
    expect(render).toHaveBeenCalled()
  })
})

describe('with renderLoading, fetchingInfo and info', () => {
  it('renders loading', () => {
    const renderLoading = jest.fn(() => <div>Loading</div>)
    const props = {
      render,
      fetchingInfo: true,
      renderLoading,
      info: { foo: 'bar' },
    }

    const tree = renderer
      .create(<Info id={1} name="colors" {...props} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
    expect(render).not.toHaveBeenCalled()
  })
})
