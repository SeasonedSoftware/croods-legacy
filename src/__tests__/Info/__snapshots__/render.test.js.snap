// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  info={undefined}
  render={[Function]}
>
  Info
</div>
`;

exports[`with renderError and infoError renders error 1`] = `
<div>
  Error: 
  foo
</div>
`;

exports[`with renderError and without infoError renders Info 1`] = `
<div
  info={undefined}
  render={[Function]}
  renderError={[Function]}
>
  Info
</div>
`;

exports[`with renderLoading and fetchingInfo renders loading correctly 1`] = `
<div>
  Loading
</div>
`;

exports[`with renderLoading and info renders info correctly 1`] = `
<div
  info={
    Object {
      "foo": "bar",
    }
  }
  render={[Function]}
  renderLoading={[Function]}
>
  Info
</div>
`;

exports[`with renderLoading, fetchingInfo and info renders loading 1`] = `
<div>
  Loading
</div>
`;

exports[`without renderLoading and with fetchingInfo renders info correctly 1`] = `
<div
  fetchingInfo={true}
  info={undefined}
  render={[Function]}
>
  Info
</div>
`;
